gNMI_Interface/gnmi_config.py

```diff
@@ -0,0 +1,241 @@
+            yield elem.name, PathListIndex(self.elems, self.index, self.value)
+        elif self.index + 1 == len(self.elems):
+            yield elem.name, new_value_index(self.value)
+        else:
+            yield elem.name, PathTreeIndex(self.elems, self.index + 1, self.value)
+
+    def new_child(self):
+        return GNMIConfigTree()
+
+
+class PathListIndex(PathIndex['PathListIndex']):
+    def config_class(self):
+        return GNMIConfigList
+
+    def subindex(self) -> PathIndex:
+        if self.index + 1 == len(self.elems):
+            return new_value_index(self.value)
+        else:
+            return PathTreeIndex(self.elems, self.index + 1, self.value)
+
+    def new_child(self):
+        return GNMIConfigList(self.elem.key)
+
+    def key_values(self) -> t.Tuple[str, ...]:
+        return tuple(v for _k, v in self.elem.key.items())
+
+
+JsonPrimitiveT = t.Union[int, str, float]
+JsonValueT = t.Union[JsonPrimitiveT, t.Dict[str, 'JsonValueT']]
+
+
+def new_value_index(value: gnmi.TypedValue):
+    if value.HasField('json_ietf_val'):
+        return new_json_index(json.loads(value.json_ietf_val))
+    else:
+        # can ListFields() be empty, is it TypedValue?
+        return PlainValueIndex(value.ListFields()[0][1])
+
+
+def new_json_index(value: JsonValueT) -> t.Union['JsonTreeIndex', 'JsonValueIndex']:
+    if isinstance(value, dict):
+        return JsonTreeIndex(value)
+    else:
+        return JsonValueIndex(value)
+
+
+class JsonValueIndex(UpdateIndex['JsonValueIndex']):
+    def __init__(self, value: JsonPrimitiveT) -> None:
+        self.value = value
+
+    def config_class(self):
+        return GNMIConfigValue
+
+    def new_child(self):
+        return GNMIConfigValue()
+
+
+class PlainValueIndex(UpdateIndex['PlainValueIndex']):
+    def __init__(self, value: t.Any) -> None:
+        self.value = value
+
+    def config_class(self):
+        return GNMIConfigValue
+
+    def new_child(self):
+        return GNMIConfigValue()
+
+
+class JsonTreeIndex(UpdateIndex['JsonTreeIndex']):
+    def __init__(self, value: dict) -> None:
+        self.value = value
+
+    def config_class(self):
+        return GNMIConfigTree
+
+    def subindices(self) -> t.Iterable[t.Tuple[str, UpdateIndex]]:
+        for name, val in self.value.items():
+            yield name, new_json_index(val)
+
+    def new_child(self):
+        return GNMIConfigTree()
+
+
+def verify_update(config: GNMIConfig, path: gnmi.Path, update: gnmi.Update,
+                  minimal_update: UpdateType) -> None:
+    if not minimal_update <= config.update(PathTreeIndex(path.elem, 0, update.val)):
+        up_str = f'{make_formatted_path(path)} = {update.val}'
+        if minimal_update == UpdateType.STRUCTURE:
+            msg = f'expected structural update, received: {up_str}'
+        else:
+            msg = f'received non-update: {up_str}'
+        raise AssertionError(msg)
+
+
+def apply_response(config: GNMIConfig, response: gnmi.SubscribeResponse,
```

Comment:
Why is this called `apply_response` and not `verify_response` ?

Reply:
The fact that the response is applied is more important than that it is verified (partly because if `minimal_update` is `NONE`, no verification is really done).